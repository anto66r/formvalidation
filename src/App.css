.error {
    border: solid red 1px;
}

.errorMsg {
    color: red;
}