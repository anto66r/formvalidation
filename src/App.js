import React from 'react';
import UserContext from './context'
import UserList from './userList'
import Form from './form';

const initUsers = [
  {name: 'Peter', lastname: 'Smith', id: 1 },
  {name: 'Paul', lastname: 'Dow', id: 2 },
  {name: 'Luke', lastname: 'Johnson', id: 3 },
  {name: 'Matthew', lastname: 'Richards', id: 4 },
]
function App() {
  const [users, setUsers] = React.useState(initUsers)
  const updateData = ({ name, lastname, email, age, id }) => {
    const index = users.findIndex(
      user => user.id === id
    );
    if (index < 0) {
      setUsers([
        ...users,
        ({ name, lastname, email, age, id: users.length + 1 }) 
      ])
    } else {
      setUsers([
        ...users.map(user => {
          if (user.id === id) {
            return ({ name, lastname, email, age, id })
          }
          return user
        })
      ])
    }
  }
  return (
    <div className="App">
      <UserContext.Provider
        value={
          {updateData}
        }
      >
        <UserList users={ users } />
        <Form />
     </UserContext.Provider>
    </div>
  );
}

export default App;
