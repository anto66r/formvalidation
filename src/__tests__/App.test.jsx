/* global expect, it, describe  */
import Enzyme, { mount, render, shallow } from "enzyme";
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import ReactDOM from 'react-dom';
import App from '../App';
import Form from '../form';

Enzyme.configure({ adapter: new Adapter() });



it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});


// Integrations!


describe('Should pass our beautiful integration tests', () => {


  it('Should create a new form', () => {
    const container = mount(<App />);
    container.find('input[name="name"]').simulate('change', { target: { name: 'name', value: 'John' } });
    container.find('input[name="lastname"]').simulate('change', { target: { name: 'lastname', value: 'Johnson' } });
    container.find('input[name="email"]').simulate('change', { target: { name: 'email', value: 'test@gmail.com' } });
    container.find('select[name="age"]').simulate('change', { target: { name: 'age', value: 21 } });
    container.find('input[type="submit"]').simulate('submit');

    expect(container.find("li").length).toBe(5);
    expect(container.find('span.user').at(4).text()).toBe('Johnson, John: 21 (test@gmail.com)')
  });


  it('Should clear new form when clicking Clear Fields button', () => {
    const container = mount(<App />);
    container.find('input[name="name"]').simulate('change', { target: { name: 'name', value: 'John' } });
    container.find('input[name="lastname"]').simulate('change', { target: { name: 'lastname', value: 'Johnson' } });
    container.find('input[name="email"]').simulate('change', { target: { name: 'email', value: 'test@gmail.com' } });
    container.find('select[name="age"]').simulate('change', { target: { name: 'age', value: 21 } });

    container.find('button.clear').simulate('click');

    expect(container.find('input[name="name"]').props().value).toBe('');
    expect(container.find('input[name="lastname"]').props().value).toBe('');
    expect(container.find('input[name="email"]').props().value).toBe('');
  });


  it('Should throw a validation error if email is wrong', () => {
    const container = mount(<App />);
    container.find('input[name="name"]').simulate('change', { target: { name: 'name', value: 'John' } });
    container.find('input[name="name"]').simulate('blur');

    container.find('input[name="lastname"]').simulate('change', { target: { name: 'lastname', value: 'Johnson' } });
    container.find('input[name="lastname"]').simulate('blur');

    container.find('input[name="email"]').simulate('change', { target: { name: 'email', value: 'blabla' } });
    container.find('input[name="email"]').simulate('blur');

    container.find('select[name="age"]').simulate('change', { target: { name: 'age', value: 21 } });
    container.find('select[name="age"]').simulate('blur');

    container.find('input[type="submit"]').simulate('submit');

    expect(container.find('.errorMsg').text()).toBe('This field must be an e-mail address')
  });


  it('Should edit a form', () => {
    const container = mount(<App />);
    container.find('button.edit').at(0).simulate('click');
    container.find('input[name="name"]').at(0).simulate('change', { target: { name: 'name',  value: 'Mark' } });
    container.find('input[name="lastname"]').at(0).simulate('change', { target: { name: 'lastname', value: 'Thompson' } });
    container.find('input[name="email"]').at(0).simulate('change', { target: { name: 'email', value: 'blabla@gmail.com' } });
    container.find('select[name="age"]').at(0).simulate('change', { target: { name: 'age', value: 22 } });
    container.find('input[type="submit"]').at(0).simulate('submit');

    expect(container.find('span.user').at(0).text()).toBe('Thompson, Mark: 22 (blabla@gmail.com)')
  });

});






