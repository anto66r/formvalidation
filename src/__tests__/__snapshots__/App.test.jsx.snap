// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should create a new form 1`] = `
<App>
  <div
    className="App"
  >
    <UserList
      users={
        Array [
          Object {
            "age": 21,
            "email": "peter@yahoo.com",
            "id": 1,
            "lastname": "Smith",
            "name": "Peter",
          },
          Object {
            "id": 2,
            "lastname": "Dow",
            "name": "Paul",
          },
          Object {
            "id": 3,
            "lastname": "Johnson",
            "name": "Luke",
          },
          Object {
            "id": 4,
            "lastname": "Richards",
            "name": "Matthew",
          },
        ]
      }
    >
      <ul>
        <ListItem
          age={21}
          email="peter@yahoo.com"
          id={1}
          key="1"
          lastname="Smith"
          name="Peter"
        >
          <li
            key="1"
          >
            Smith, Peter: 21 (peter@yahoo.com)
            <button
              onClick={[Function]}
              type="button"
            >
              Edit
            </button>
          </li>
        </ListItem>
        <ListItem
          id={2}
          key="2"
          lastname="Dow"
          name="Paul"
        >
          <li
            key="2"
          >
            Dow, Paul: n/a (n/a)
            <button
              onClick={[Function]}
              type="button"
            >
              Edit
            </button>
          </li>
        </ListItem>
        <ListItem
          id={3}
          key="3"
          lastname="Johnson"
          name="Luke"
        >
          <li
            key="3"
          >
            Johnson, Luke: n/a (n/a)
            <button
              onClick={[Function]}
              type="button"
            >
              Edit
            </button>
          </li>
        </ListItem>
        <ListItem
          id={4}
          key="4"
          lastname="Richards"
          name="Matthew"
        >
          <li
            key="4"
          >
            Richards, Matthew: n/a (n/a)
            <button
              onClick={[Function]}
              type="button"
            >
              Edit
            </button>
          </li>
        </ListItem>
      </ul>
    </UserList>
    <_default>
      <form
        onSubmit={[Function]}
      >
        <_default
          handleBlur={[Function]}
          handleChange={[Function]}
          label="Name"
          name="name"
        >
          <label
            htmlFor="name"
          >
            Name
            :
            <input
              name="name"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />
          </label>
        </_default>
        <_default
          handleBlur={[Function]}
          handleChange={[Function]}
          label="Last name"
          name="lastname"
        >
          <label
            htmlFor="lastname"
          >
            Last name
            :
            <input
              name="lastname"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />
          </label>
        </_default>
        <_default
          handleBlur={[Function]}
          handleChange={[Function]}
          label="E-mail"
          name="email"
        >
          <label
            htmlFor="email"
          >
            E-mail
            :
            <input
              name="email"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />
          </label>
        </_default>
        <_default
          handleBlur={[Function]}
          handleChange={[Function]}
          label="Age"
          name="age"
          values={
            Array [
              "",
              19,
              20,
              21,
              22,
            ]
          }
        >
          <label
            htmlFor="age"
          >
            Age
            :
            <select
              name="age"
              onChange={[Function]}
            >
              <option
                key=""
              />
              <option
                key="19"
              >
                19
              </option>
              <option
                key="20"
              >
                20
              </option>
              <option
                key="21"
              >
                21
              </option>
              <option
                key="22"
              >
                22
              </option>
            </select>
          </label>
        </_default>
        <input
          disabled={false}
          type="submit"
        />
        <button
          onClick={[Function]}
          type="button"
        >
          Clear fields
        </button>
      </form>
    </_default>
  </div>
</App>
`;

exports[`we change first render 1`] = `
<form
  onSubmit={[Function]}
>
  <_default
    handleBlur={[Function]}
    handleChange={[Function]}
    label="Name"
    name="name"
  />
  <_default
    handleBlur={[Function]}
    handleChange={[Function]}
    label="Last name"
    name="lastname"
  />
  <_default
    handleBlur={[Function]}
    handleChange={[Function]}
    label="E-mail"
    name="email"
  />
  <_default
    handleBlur={[Function]}
    handleChange={[Function]}
    label="Age"
    name="age"
    values={
      Array [
        "",
        19,
        20,
        21,
        22,
      ]
    }
  />
  <input
    disabled={true}
    type="submit"
  />
  <button
    onClick={[Function]}
    type="button"
  >
    Clear fields
  </button>
</form>
`;
